### PR TITLE
feat: Add nursing tasks

### PR DESCRIPTION
**Pull request by ChillarAnand**
Docs: https://docs.erpnext.com/docs/v13/user/manual/en/nursing_tasks

From checking vitals to ensuring medications are administered to a patient in a timely manner, nursing tasks are essential part of healthcare workflows. Frappe Healthcare ensures nursing tasks are transitioned correctly through out the workflows.

Creating nursing checklist template.

Screenshot 2021-11-08 at 11 11 49

Link nursing checklist to lab test template.

Screenshot 2021-11-08 at 11 13 04

Create lab test from template and it should create nursing tasks.

Screenshot 2021-11-08 at 11 13 20

Users cant submit lab test before completing nursing tasks.

Screenshot 2021-11-08 at 11 13 39

To see linked nursing tasks click nursing tasks button.

Screenshot 2021-11-08 at 11 13 48

To modify nursing task on a case to case basis, click on `modify nursing tasks` and select required tasks.

Screenshot 2021-11-25 at 18 37 50

Screenshot 2021-11-25 at 18 38 02

### CODE REVIEW

**Comment by ruchamahabal** [comment]
@ChillarAnand Can you update the PR description with why the feature was added and the flow of creation?

**Comment by ChillarAnand** [comment]
> @ChillarAnand Can you update the PR description with why the feature was added and the flow of creation?

Added docs for the same at https://docs.erpnext.com/docs/v13/user/manual/en/nursing_tasks. @ruchamahabal

File: healthcare/healthcare/doctype/nursing_checklist_template/nursing_checklist_template.json [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  "department",
+  "tasks_section",
+  "tasks"
+ ],
+ "fields": [
+  {
+   "fieldname": "title",
+   "fieldtype": "Data",
+   "label": "Title",
+   "unique": 1
```
Comment (ruchamahabal) [inline healthcare/healthcare/doctype/nursing_checklist_template/nursing_checklist_template.json:23]: Title should be mandatory

File: healthcare/healthcare/doctype/healthcare_activity/healthcare_activity.json [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+   "read": 1,
+   "report": 1,
+   "role": "System Manager",
+   "share": 1,
+   "write": 1
+  }
+ ],
+ "sort_field": "modified",
+ "sort_order": "DESC",
+ "track_changes": 1
```
Comment (ruchamahabal) [inline healthcare/healthcare/doctype/healthcare_activity/healthcare_activity.json:87]: Enable Quick Entry for this doctype so that it can be created from any view without navigating to that form.

File: healthcare/healthcare/doctype/nursing_checklist_template/nursing_checklist_template.json [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+   "options": "Medical Department"
+  },
+  {
+   "fieldname": "tasks_section",
+   "fieldtype": "Section Break",
+   "label": "Tasks"
+  },
+  {
+   "fieldname": "tasks",
+   "fieldtype": "Table",
```
Comment (ruchamahabal) [inline healthcare/healthcare/doctype/nursing_checklist_template/nursing_checklist_template.json]: Redundant labeling for table and section? Just keep the section label

image

File: healthcare/healthcare/doctype/healthcare_activity/healthcare_activity.json (hunk 2 above)
Comment (ruchamahabal) [reply healthcare/healthcare/doctype/healthcare_activity/healthcare_activity.json:87]: Set auto naming for this doctype based on the activity field
image

File: healthcare/healthcare/doctype/nursing_checklist_template/nursing_checklist_template.json (hunk 3 above)
Comment (ruchamahabal) [reply healthcare/healthcare/doctype/nursing_checklist_template/nursing_checklist_template.json]: Layout in Child Table and a few pointers:

image

- This table should be mandatory
- Description should not be in grid view since it can be opened for more context. Most nurses would be familiar with the title of the activity
- This field should be a part of Activity doctype instead since it can be reused in multiple templates and users won't keep filling it up for each. Fetch desc from the activity and let the user edit it if need be.
- **Starts On** needs to be an Int field. It is meant for the day on which this should be done relative to the date of the transaction. Since this template will be used in multiple transactions same date won't be applicable for all. Check [Employee Onboarding Template](https://docs.erpnext.com/docs/v13/user/manual/en/human-resources/employee-onboarding#31-employee-onboarding-template) for context.
- Following fields should be shown in Grid View instead:
	- Activity
	- Type
	- Starts On
	- Duration
	- Mandatory
	- Role
- Add a column break in the child table layout to unclutter the fields.
- Add more context for the Task DocType field. Labeling should be Reference Document Type. You can add a description for this field: Reference Document to be Validated (eg: If reference is Vital Signs system will validate whether vital signs were actually captured during the nursing task)

File: healthcare/healthcare/doctype/lab_test/lab_test.json [hunk 4]
```diff
@@ -555,14 +556,21 @@
...
+   "fetch_from": "template.nursing_checklist_template",
+   "fieldname": "modify_nursing_tasks",
+   "fieldtype": "Button",
+   "label": "Modify Nursing Tasks "
   }
  ],
  "is_submittable": 1,
  "links": [],
- "modified": "2020-11-30 11:04:17.195848",
+ "modified": "2021-11-29 06:16:23.354524",
```
Comment (ruchamahabal) [inline healthcare/healthcare/doctype/lab_test/lab_test.json]: For some reason, the patient is getting reset in Lab Test. Looks like it's trying to fetch from some field.

Kapture 2022-01-12 at 13 06 55

File: healthcare/healthcare/doctype/lab_test/lab_test.py [hunk 5]
```diff
@@ -19,6 +20,8 @@ def on_submit(self):
 		self.validate_result_values()
 		self.db_set('submitted_date', getdate())
 		self.db_set('status', 'Completed')
+		from healthcare.healthcare.utils import validate_nursing_tasks
```
Comment (ruchamahabal) [inline healthcare/healthcare/doctype/lab_test/lab_test.py]: We can keep all the imports together?

File: healthcare/healthcare/doctype/lab_test/lab_test.js [hunk 6]
```diff
@@ -57,9 +57,49 @@ frappe.ui.form.on('Lab Test', {
 			});
 		}
 
-	}
+		frm.add_custom_button(__("Nursing Tasks"), function () {
+				frappe.set_route("List", "Nursing Task", {"reference_name": frm.doc.name});
+		});
```
Comment (ruchamahabal) [inline healthcare/healthcare/doctype/lab_test/lab_test.js]: This button should not be shown before saving:

image

Also, club it under a parent button View -> Nursing Tasks. Since just "Nursing Tasks" is ambiguous.

Comment (ruchamahabal) [inline healthcare/healthcare/doctype/lab_test/lab_test.js]:
```suggestion
		frm.add_custom_button(__("Nursing Tasks"), function () {
			frappe.set_route("List", "Nursing Task", {"reference_name": frm.doc.name});
		});
```

Over indented?

File: healthcare/healthcare/doctype/lab_test/lab_test.js [hunk 7]
```diff
@@ -57,9 +57,49 @@ frappe.ui.form.on('Lab Test', {
...
+							method: 'modify_nursing_tasks',
+							doc: frm.doc,
+							args: nursing_tasks,
+						});
+						cur_dialog.hide();
+					}
+				});
+			}
+		});
+	},
```
Comment (ruchamahabal) [inline healthcare/healthcare/doctype/lab_test/lab_test.js]: - Mixed tabs and spaces:
- Also try to use uniform quote styles throughout a single file. Applies to all files

image

File: healthcare/healthcare/doctype/lab_test/lab_test.py [hunk 8]
```diff
@@ -67,6 +76,34 @@ def validate_result_values(self):
 					frappe.throw(_('Row #{0}: Please enter the result value for {1}').format(
 						item.idx, frappe.bold(item.lab_test_particulars)), title=_('Mandatory Results'))
 
+	@frappe.whitelist()
+	def modify_nursing_tasks(self, nursing_tasks):
```
Comment (ruchamahabal) [inline healthcare/healthcare/doctype/lab_test/lab_test.py]: This should first save the document before creating the tasks:

image

File: healthcare/healthcare/doctype/lab_test/lab_test.py [hunk 9]
```diff
@@ -67,6 +76,34 @@ def validate_result_values(self):
...
+		template_tasks = frappe.get_list(
+			'Nursing Checklist Template Task',
+			filters=self.get_nursing_task_filters(self.template),
+			fields=['name', 'activity', 'mandatory', 'task_doctype'],
+		)
```
Comment (ruchamahabal) [inline healthcare/healthcare/doctype/lab_test/lab_test.py]:
```suggestion
		template_tasks = frappe.get_all(
			'Nursing Checklist Template Task',
			filters=self.get_nursing_task_filters(self.template),
			fields=['name', 'activity', 'mandatory', 'task_doctype'],
		)
```

We should not use get_list on the child table since it checks for perms.

File: healthcare/healthcare/doctype/lab_test/lab_test.py [hunk 10]
```diff
@@ -67,6 +76,34 @@ def validate_result_values(self):
...
+		existing_tasks = frappe.get_list(
+			'Nursing Task',
+			filters={'reference_name': self.name, 'docstatus': 0},
+		)
```
Comment (ruchamahabal) [inline healthcare/healthcare/doctype/lab_test/lab_test.py]:
```suggestion
		existing_tasks = frappe.get_all(
			'Nursing Task',
			filters={'reference_name': self.name, 'docstatus': 0},
		)
```

Again since this is happening in the backend safer to use `get_all`? The user will get unnecessary perm error.

File: healthcare/healthcare/doctype/lab_test/lab_test.py [hunk 11]
```diff
@@ -67,6 +76,34 @@ def validate_result_values(self):
...
+
+		if not existing_tasks:
+			NursingTask.create_nursing_tasks('Lab Test', self.name, template_tasks)
+			return
```
Comment (ruchamahabal) [inline healthcare/healthcare/doctype/lab_test/lab_test.py]: Why is this needed? Creation should always happen as per selection right, which is happening in lines 98-99? Irrespective of whether any task exists or not.

File: healthcare/healthcare/doctype/lab_test/lab_test.py (hunk 8 above)
Comment (ruchamahabal) [reply healthcare/healthcare/doctype/lab_test/lab_test.py]: Now that this is in the picture some issues I see here:
- users might miss this button during creation. We should allow it after saving too, delete the task if it's modified.
- There is no version log maintained at the bottom of the doc since this modification is stateless. Should we insert one for context?

File: healthcare/healthcare/doctype/lab_test/lab_test.json [hunk 12]
```diff
@@ -555,14 +556,21 @@
    "label": "Organism Test Result",
    "options": "Organism Test Result",
    "print_hide": 1
+  },
+  {
+   "fetch_from": "template.nursing_checklist_template",
+   "fieldname": "modify_nursing_tasks",
+   "fieldtype": "Button",
+   "label": "Modify Nursing Tasks "
```
Comment (ruchamahabal) [inline healthcare/healthcare/doctype/lab_test/lab_test.json]: Any reason for keeping this button only in Lab Test?

File: healthcare/healthcare/doctype/clinical_procedure/clinical_procedure.js [hunk 13]
```diff
@@ -131,6 +131,10 @@ frappe.ui.form.on('Clinical Procedure', {
 				}).addClass("btn-primary");
 			}
 		}
+
+		frm.add_custom_button(__("Nursing Tasks"), function () {
+				frappe.set_route("List", "Nursing Task", {"reference_name": frm.doc.name});
+		});
```
Comment (ruchamahabal) [inline healthcare/healthcare/doctype/clinical_procedure/clinical_procedure.js]: This button should not be shown before saving:
Also, club it under a parent button View -> Nursing Tasks. Since just "Nursing Tasks" is ambiguous.

File: healthcare/healthcare/doctype/clinical_procedure/clinical_procedure.py [hunk 14]
```diff
@@ -45,6 +47,17 @@ def after_insert(self):
 			frappe.db.set_value('Clinical Procedure', self.name, 'sample', sample_collection.name)
 		self.reload()
 
+		if self.procedure_template:
+			cp_template = frappe.get_doc('Clinical Procedure Template', self.procedure_template)
+			templates = [cp_template.pre_op_nursing_checklist_template, cp_template.post_op_nursing_checklist_template]
+			for template in templates:
+				if not template:
+					continue
+				NursingTask.create_nursing_tasks_from_template(template, 'Clinical Procedure', self.name)
```
Comment (ruchamahabal) [inline healthcare/healthcare/doctype/clinical_procedure/clinical_procedure.py]:
```suggestion
			cp_templates = frappe.db.get_value('Clinical Procedure Template', self.procedure_template, ['pre_op_nursing_checklist_template', 'post_op_nursing_checklist_template'], as_dict=1)
			NursingTask.create_nursing_tasks_from_template(cp_templates.pre_op_nursing_checklist_template, 'Clinical Procedure', self.name)
			NursingTask.create_nursing_tasks_from_template(cp.templates.post_op_nursing_checklist_template, 'Clinical Procedure', self.name)
```

- We can avoid `get_doc` here. `get_doc` is always expensive than `db.get_value`.
- Unnecessary for loop. Just call the method twice.

File: healthcare/healthcare/doctype/healthcare.py [hunk 15]
```diff
@@ -1,6 +1,12 @@
...
 
+from frappe.model.document import Document
+
 
 def get_data():
 
 	return []
+
+
+class PatientDocument(Document):
```
Comment (ruchamahabal) [inline healthcare/healthcare/doctype/healthcare.py]: What's this for?

File: healthcare/healthcare/doctype/healthcare_settings/healthcare_settings.js [hunk 16]
```diff
@@ -56,6 +56,11 @@ frappe.tour['Healthcare Settings'] = [
 		title: __('Automate Appointment Invoicing'),
 		description: __('Checking this will automatically create a Sales Invoice whenever an appointment is booked for a Patient.')
 	},
+	{
+		fieldname: 'validate_nursing_checklists',
+		title: __('validate_nursing_checklists'),
+		description: __('Checking this will automatically create a Sales Invoice whenever an appointment is booked for a Patient.')
```
Comment (ruchamahabal) [inline healthcare/healthcare/doctype/healthcare_settings/healthcare_settings.js]: Check Description

File: healthcare/healthcare/doctype/inpatient_record/inpatient_record.js [hunk 17]
```diff
@@ -34,6 +34,10 @@ frappe.ui.form.on('Inpatient Record', {
 		} else {
 			frm.set_df_property('btn_transfer', 'hidden', 0);
 		}
+
+		frm.add_custom_button(__("Nursing Tasks"), function () {
+				frappe.set_route("List", "Nursing Task", {"reference_name": frm.doc.name});
+		});
```
Comment (ruchamahabal) [inline healthcare/healthcare/doctype/inpatient_record/inpatient_record.js]: This button should not be shown before saving:
Also, club it under a parent button View -> Nursing Tasks. Since just "Nursing Tasks" is ambiguous.

File: healthcare/healthcare/doctype/inpatient_record/inpatient_record.json [hunk 18]
```diff
@@ -408,14 +410,27 @@
...
+  {
+   "fieldname": "discharge_nursing_checklist_template",
+   "fieldtype": "Link",
+   "label": "Discharge Nursing Checklist Template",
+   "options": "Nursing Checklist Template"
+  },
+  {
+   "fieldname": "admission_nursing_checklist_template",
+   "fieldtype": "Link",
+   "label": "Nursing Checklist Template",
```
Comment (ruchamahabal) [inline healthcare/healthcare/doctype/inpatient_record/inpatient_record.json]:
```suggestion
   "label": "Admission Nursing Checklist Template",
```

File: healthcare/healthcare/doctype/lab_test/lab_test.py [hunk 19]
```diff
@@ -67,6 +76,34 @@ def validate_result_values(self):
...
+		for task in existing_tasks:
+			frappe.delete_doc('Nursing Task', task['name'])
+
+		selected_tasks = [task for task in template_tasks if task['name'] in nursing_tasks]
+		NursingTask.create_nursing_tasks('Lab Test', self.name, selected_tasks)
+
+	@frappe.whitelist()
+	def get_nursing_task_filters(self, lab_test_template):
+		lab_test_template = frappe.get_doc('Lab Test Template', lab_test_template)
+		filters = {'parent': lab_test_template.nursing_checklist_template}
```
Comment (ruchamahabal) [inline healthcare/healthcare/doctype/lab_test/lab_test.py]:
```suggestion
		nursing_checklist_template = frappe.db.get_value('Lab Test Template', lab_test_template, 'nursing_checklist_template')
		filters = {'parent': nursing_checklist_template}
```

Also, do we need this whitelisted method? It's just called from a single function, just compute it there itself.

File: healthcare/healthcare/doctype/nursing_checklist_template/nursing_checklist_template.json [hunk 20]
```diff
@@ -0,0 +1,65 @@
...
+   "fieldname": "title",
+   "fieldtype": "Data",
+   "label": "Title",
+   "unique": 1
+  },
+  {
+   "fieldname": "department",
+   "fieldtype": "Link",
+   "label": "Department",
+   "options": "Medical Department"
```
Comment (ruchamahabal) [inline healthcare/healthcare/doctype/nursing_checklist_template/nursing_checklist_template.json:31]: Add Department to List View and Filter View:

image

File: healthcare/healthcare/doctype/nursing_task/nursing_task.json [hunk 21]
```diff
@@ -0,0 +1,157 @@
...
+   "read": 1,
+   "report": 1,
+   "role": "System Manager",
+   "share": 1,
+   "write": 1
+  }
+ ],
+ "sort_field": "modified",
+ "sort_order": "DESC",
+ "track_changes": 1
```
Comment (ruchamahabal) [inline healthcare/healthcare/doctype/nursing_task/nursing_task.json:383]: image

- Set title for this doctype, maybe activity
- Add more fields to List View and Filter

File: healthcare/healthcare/doctype/nursing_task/nursing_task.json [hunk 22]
```diff
@@ -0,0 +1,157 @@
...
+   "fieldtype": "Date",
+   "label": "To Date"
+  },
+  {
+   "fieldname": "status",
+   "fieldtype": "Select",
+   "in_list_view": 1,
+   "label": "Status",
+   "options": "Requested\nDraft\nReceived",
+   "reqd": 1
```
Comment (ruchamahabal) [inline healthcare/healthcare/doctype/nursing_task/nursing_task.json]: Instead of these in reference to [https://www.hl7.org/fhir/valueset-task-status.html](https://www.hl7.org/fhir/valueset-task-status.html) should we have:

- Draft
- Cancelled
- Requested
- In Progress
- On Hold
- Failed
- Completed

File: healthcare/healthcare/utils.py [hunk 23]
```diff
@@ -814,4 +814,18 @@ def before_tests():
...
 			"domains"           : ["Healthcare"],
 		})
+
 		setup_healthcare()
+
+
+def validate_nursing_tasks(document):
+	healthcare_settings = frappe.get_single("Healthcare Settings")
+	if not healthcare_settings.validate_nursing_checklists:
+		return True
```
Comment (ruchamahabal) [inline healthcare/healthcare/utils.py]:
```suggestion
	if not frappe.db.get_single_value("Healthcare Settings", "validate_nursing_checklists")
		return True
```

File: healthcare/healthcare/utils.py [hunk 24]
```diff
@@ -814,4 +814,18 @@ def before_tests():
...
+
+	filters = {'reference_name': document.name, 'docstatus': 0}
+	tasks = frappe.get_all('Nursing Task', filters=filters)
+	if not tasks:
+		return True
+
+	frappe.throw('Please complete related nursing tasks before submission')
```
Comment (ruchamahabal) [inline healthcare/healthcare/utils.py]:
```suggestion
	frappe.throw(_('Please complete linked Nursing Tasks before submission'))
```

Comment (ruchamahabal) [reply healthcare/healthcare/utils.py]: Also, what's the purpose of marking nursing tasks as Mandatory? We should ignore validating non-mandatory tasks here. Also, while making the checklist, let mandatory be set by default. User can uncheck for whichever tasks they don't want mandatory.

File: healthcare/healthcare/doctype/nursing_task/nursing_task.json [hunk 25]
```diff
@@ -0,0 +1,157 @@
...
+   "fieldtype": "Link",
+   "in_list_view": 1,
+   "label": "Activity",
+   "options": "Healthcare Activity",
+   "reqd": 1
+  },
+  {
+   "fieldname": "description",
+   "fieldtype": "Data",
+   "label": "Description"
```
Comment (ruchamahabal) [inline healthcare/healthcare/doctype/nursing_task/nursing_task.json]: - The description should be read-only and the same data type as in the checklist/activity. Set it from Checklist desc
- **Mandatory** checkbox should be **Set Only Once**?
- Add role field in Nursing Task since we set it in the template. If Role is set, apply a filter on the user field for users with that role else keep it open-ended.
- Rearrange fields
	- LHS: Activity, User, Description
	- RHS: Status, From Date, To Date
- Reference Doctype --> Reference Document Type
- Task DocType --> Task Document Type

image

The whole task reference concept is not very evident from the design. Need to rethink the purpose, utility, and usage. Should be easy for the user to understand.

File: healthcare/healthcare/doctype/therapy_plan/therapy_plan.py [hunk 26]
```diff
@@ -48,6 +53,14 @@ def set_therapy_details_from_template(self):
 			})
 		return self
 
+	def after_insert(self):
+		if self.therapy_plan_template:
+			therapy_template = frappe.get_doc('Therapy Plan Template', self.therapy_plan_template)
```
Comment (ruchamahabal) [inline healthcare/healthcare/doctype/therapy_plan/therapy_plan.py]: Replace with `get_value` as pointed out in the comments earlier.

File: healthcare/healthcare/doctype/nursing_task/nursing_task.py [hunk 27]
```diff
@@ -0,0 +1,42 @@
+# Copyright (c) 2021, healthcare and contributors
+# For license information, please see license.txt
+import frappe
+from frappe.model.document import Document
+
+
+class NursingTask(Document):
+	@classmethod
+	def create_nursing_tasks_from_template(cls, template_name=None, dt=None, dn=None, append=False):
+		tasks = frappe.get_list(
```
Comment (ruchamahabal) [inline healthcare/healthcare/doctype/nursing_task/nursing_task.py]:
```suggestion
		tasks = frappe.get_all(
```

File: healthcare/healthcare/doctype/nursing_task/nursing_task.py [hunk 28]
```diff
@@ -0,0 +1,42 @@
...
+			fields=['*'],
+		)
+		if append:
+			NursingTask.create_nursing_tasks(dt, dn, tasks)
+		else:
+			filters = {
+				'reference_doctype': dt,
+				'reference_name': dn,
+			}
+			existing_tasks = frappe.get_list('Nursing Task', filters=filters)
```
Comment (ruchamahabal) [inline healthcare/healthcare/doctype/nursing_task/nursing_task.py]:
```suggestion
			existing_tasks = frappe.get_all('Nursing Task', filters=filters)
```

File: healthcare/healthcare/doctype/nursing_task/test_nursing_task.py [hunk 29]
```diff
@@ -0,0 +1,42 @@
...
+
+		template.append('tasks', task)
+		template.save()
+
+		self.lab_template = create_lab_test_template()
+		self.lab_template.nursing_checklist_template = template.name
+		self.lab_template.save()
+		create_lab_test(self.lab_template)
+
+		self.assertEqual(frappe.db.count('Nursing Task'), task_count + 1)
```
Comment (ruchamahabal) [inline healthcare/healthcare/doctype/nursing_task/test_nursing_task.py]: Increase test coverage:

- Modifying checklists
- Discharge / Pre op / Post op / Admission checklists working properly

File: healthcare/healthcare/doctype/therapy_session/therapy_session.py [hunk 30]
```diff
@@ -58,6 +61,14 @@ def on_cancel(self):
 
 		self.update_sessions_count_in_therapy_plan(on_cancel=True)
 
+	def after_insert(self):
+		therapy_plan = frappe.get_doc('Therapy Plan', self.therapy_plan)
+		if therapy_plan.therapy_plan_template:
+			plan_template = frappe.get_doc('Therapy Plan Template', therapy_plan.therapy_plan_template)
+			template = plan_template.nursing_checklist_template
```
Comment (ruchamahabal) [inline healthcare/healthcare/doctype/therapy_session/therapy_session.py]: For therapy sessions, checklists should be linked to Therapy Type instead of Plan. Also, avoid `get_doc`

File: healthcare/healthcare/doctype/nursing_task/nursing_task.py [hunk 31]
```diff
@@ -0,0 +1,42 @@
...
+	def create_nursing_tasks(cls, dt=None, dn=None, tasks=None):
+		for task in tasks:
+			options = {
+				'doctype': 'Nursing Task',
+				'status': 'Requested',
+				'reference_doctype': dt,
+				'reference_name': dn,
+				'activity': task.activity,
+				'mandatory': task.mandatory,
+				'task_doctype': task.task_doctype,
```
Comment (ruchamahabal) [inline healthcare/healthcare/doctype/nursing_task/nursing_task.py]: No mapping of dates from template to task?

The From Date should be set relative to Reference Doctype's transaction date. And End Date should be calculated as From Date + Duration

Check how [Onboarding Template](https://docs.erpnext.com/docs/v13/user/manual/en/human-resources/employee-onboarding#31-employee-onboarding-template) works in ERPNext.

**Review by ruchamahabal — CHANGES_REQUESTED** [review]

File: healthcare/healthcare/utils.py (hunk 24 above)
Comment (ruchamahabal) [reply healthcare/healthcare/utils.py]: Add this mandatory check to the multi-select dialog as well

**Comment by ruchamahabal** [comment]
General Pointers:

- Title/Naming for all doctypes
- List View/Filter defaults for all doctypes
- Document States and color mapping
- Default Roles for all doctypes (currently all are just accessible to System Managers)
- Connections/Document Links should be configured for all doctypes for easy navigation and user context
- Try to keep uniform quote styles everywhere or at least throughout a single file.
- Mixed tabs/spaces in some files. Stick to tabs as per our current codebase.
- Check search fields for every doctype
- IMO, since this is applicable for every type of template, tests should cover it for all doctypes: Clinical Procedure, Admission, Discharge, Therapy Session, and not just Lab Test

File: healthcare/healthcare/doctype/lab_test/lab_test.py (hunk 5 above)
Comment (akurungadam) [reply healthcare/healthcare/doctype/lab_test/lab_test.py]: this is to get around a circular import, will fix this as part of the lab module redesign

File: healthcare/healthcare/doctype/lab_test/lab_test.js (hunk 6 above)
Comment (akurungadam) [reply healthcare/healthcare/doctype/lab_test/lab_test.js]: moved the custom buttons to Connections

File: healthcare/healthcare/doctype/lab_test/lab_test.py (hunk 8 above)
Comment (akurungadam) [reply healthcare/healthcare/doctype/lab_test/lab_test.py]: Not too sure if we should keep the `Modify Nursing Tasks`. Maybe bring this in based on request / feedback?

File: healthcare/healthcare/doctype/healthcare.py (hunk 15 above)
Comment (akurungadam) [reply healthcare/healthcare/doctype/healthcare.py]: Removed this for now, Anand you'll add back as needed?

File: healthcare/healthcare/doctype/nursing_task/nursing_task.json (hunk 21 above)
Comment (akurungadam) [reply healthcare/healthcare/doctype/nursing_task/nursing_task.json:383]: Screen Shot 2022-01-18 at 1 26 55 PM

File: healthcare/healthcare/doctype/nursing_task/nursing_task.json (hunk 22 above)
Comment (akurungadam) [reply healthcare/healthcare/doctype/nursing_task/nursing_task.json]: added some more optional statuses (Ready, Approved, Rejected, Entered in Error etc.)

File: healthcare/healthcare/utils.py (hunk 24 above)
Comment (akurungadam) [reply healthcare/healthcare/utils.py]: Checking for 'Completed' status now, Nursing Tasks are now submitted on status 'Requested'

File: healthcare/healthcare/doctype/nursing_task/nursing_task.py (hunk 31 above)
Comment (akurungadam) [reply healthcare/healthcare/doctype/nursing_task/nursing_task.py]: Added a time_offset field to set the requested_start_time

File: healthcare/healthcare/doctype/lab_test/lab_test.json (hunk 4 above)
Comment (akurungadam) [reply healthcare/healthcare/doctype/lab_test/lab_test.json]: couldn't see this happening, could you please check once again?

File: healthcare/healthcare/doctype/healthcare_activity/healthcare_activity.json (hunk 2 above)
Comment (akurungadam) [reply healthcare/healthcare/doctype/healthcare_activity/healthcare_activity.json:87]: added title as name field and enabled quick entry

**Review by ruchamahabal — APPROVED** [review]
